### PR TITLE
toke.c: fix =cut detection

### LINKED ISSUE #22759: POD section unexpectedly stops at =cut2 (or =cute in string eval)
**Description**
The perl parser is supposed to skip over embedded POD sections. The end of a POD section is marked by a `=cut` command. However, not every line that starts with `=cut` is a `=cut` command. For example: `=cute`, `=cut2`, `=cut_dry`.

The perl parser does not correctly recognize these and, in the case of `=cute`, behaves differently depending on whether we are in string eval or not.

**Steps to Reproduce**
```perl
use strict;
use warnings;

print "1..9\n";

my $test = 1;

my $foo;
$foo = "";

=pod

=cute
$foo = "not ";

=pod

=cut

print $foo, "ok $test - =cute does not end POD\n";
$test++;

$foo = "";

=pod

=cut2
$foo = "not ";

=pod

=cut

print $foo, "ok $test - =cut2 does not end POD\n";
$test++;

$foo = "";

=pod

=cut_
$foo = "not ";

=pod

=cut

print $foo, "ok $test - =cut_ does not end POD\n";
$test++;

$foo = "not ";

=pod

=cut$cene
$foo = "";

=pod

=cut

print $foo, "ok $test - =cut\$cene ends POD\n";
$test++;

# Same as above, but in string eval.

eval q{
$foo = "";

=pod

=cute
$foo = "not ";

=pod

=cut

print $foo, "ok $test - =cute does not end POD\n";
$test++;

$foo = "";

=pod

=cut2
$foo = "not ";

=pod

=cut

print $foo, "ok $test - =cut2 does not end POD\n";
$test++;

$foo = "";

=pod

=cut_
$foo = "not ";

=pod

=cut

print $foo, "ok $test - =cut_ does not end POD\n";
$test++;

$foo = "not ";

=pod

=cut$cene
$foo = "";

=pod

=cut

print $foo, "ok $test - =cut\$cene ends POD\n";
$test++;
};

print $@ eq "" ? "" : "not ", "ok $test - did not throw an error\n# $@\n";
__END__
```
Output:
```
1..9
ok 1 - =cute does not end POD
not ok 2 - =cut2 does not end POD
not ok 3 - =cut_ does not end POD
ok 4 - =cut$cene ends POD
not ok 5 - =cute does not end POD
not ok 6 - =cut2 does not end POD
not ok 7 - =cut_ does not end POD
ok 8 - =cut$cene ends POD
ok 9 - did not throw an error
# 
```

**Expected behavior**
```
1..9
ok 1 - =cute does not end POD
ok 2 - =cut2 does not end POD
ok 3 - =cut_ does not end POD
ok 4 - =cut$cene ends POD
ok 5 - =cute does not end POD
ok 6 - =cut2 does not end POD
ok 7 - =cut_ does not end POD
ok 8 - =cut$cene ends POD
ok 9 - did not throw an error
# 
```

**Perl configuration**

This is with v5.40.0, but it should affect all perl versions (except before 5.8.9 the `=cute` test will probably fail without eval, too).

### PR DESCRIPTION
**Pull request by mauke**
A line starting with `=cut` is only a cut directive if the next character is not a word character. Checking for isALPHA is insufficient because POD directives can contain digits (e.g. `=head1`, `=head2`).

Add the same check to the string eval case, where it was missing entirely.

Fixes #22759.

---------------------------------------------------------------------------------
* This set of changes requires a perldelta entry, and it is included.

### CODE REVIEW

File: toke.c [hunk 1]
```diff
@@ -9372,7 +9372,8 @@ yyl_try(pTHX_ char *s)
                     while (s < d) {
                         if (*s++ == '\n') {
                             incline(s, PL_bufend);
-                            if (memBEGINs(s, (STRLEN) (PL_bufend - s), "=cut"))
+                            if (memBEGINPs(s, (STRLEN) (PL_bufend - s), "=cut")
+                                && !isWORDCHAR_A(s[4]))
```
Comment (leonerd) [inline toke.c]: Do we know that `s[4]` is valid at this point? Is there a length guard somewhere?

Comment (mauke) [reply toke.c]: That's what the `BEGINP` (P = "proper") macro is for: It checks for a proper prefix, i.e. `s` must have at least 5 characters.

File: toke.c [hunk 2]
```diff
@@ -7349,7 +7349,7 @@ yyl_fake_eof(pTHX_ U32 fake_eof, bool bof, char *s)
         if (PL_parser->in_pod) {
             /* Incest with pod. */
             if (    memBEGINPs(s, (STRLEN) (PL_bufend - s), "=cut")
-                && !isALPHA(s[4]))
+                && !isWORDCHAR_A(s[4]))
```
Comment (khwilliamson) [inline toke.c]: This should probably instead be `!isIDCONT_A` for Unicode-aware use.  The two macros are equivalent in the ASCII range, but at some point we may want to expand beyond that range, and the proper semantics are IDCONT

Comment (mauke) [reply toke.c]: Done.

**Review by khwilliamson — APPROVED** [review]

**Pull request merged**